missingdeps tool, take 2

ilor opened the pull request:
This is a tool that detects some classes of random build failure causes around generated files (in Chrome, these are usually generated headers for C/C++). It takes several seconds to run, (more if the optional scan for files with no generator is used), and allows fixing a significant number of build system issues before they happen to randomly break someone's build.

The check relies on the correctness of depfile information, and on having all generator outputs listed. These things are usually easier to get right than proper deps. There's some overlap between this and gn check, but unlike gn check, this checks the actual ninja build graph and sees all issues.

This is a redo of https://github.com/ninja-build/ninja/pull/1031, it's been a while so I think it maybe better to start from scratch.

atetubou commented on src/ninja.cc:
```diff
@@ -505,6 +507,220 @@ int NinjaMain::ToolDeps(const Options* options, int argc, char** argv) {
+    }
+  }
+
+  if (!deplog_nodes_without_generator.empty()) {
```
Feel redundant.

ilor replied:
I think this is also obsolete after code move to src/missing_deps.cc and -g tool removal

atetubou commented on src/ninja.cc:
```diff
@@ -505,6 +507,220 @@ int NinjaMain::ToolDeps(const Options* options, int argc, char** argv) {
+    missing_dep_path_count_ += missing_deps_rule_names.size();
+    for (set<string>::iterator re = missing_deps_rule_names.begin();
+         re != missing_deps_rule_names.end(); ++re) {
+      //printf("Error: %s lacks a dep path to %s\n", node->path().c_str(),
```
remove or uncomment?

ilor replied:
obsolete after code move to src/missing_deps.cc (with cleanups like this)

ilor commented:
Thanks for the sugesstions so far. I'm also uploading example output of this tool, when run on a fairly recent chrome-linux build for the chrome target: [missingdeps-chrome-output.gz](https://github.com/ninja-build/ninja/files/1318295/missingdeps-chrome-output.gz)

The summary is:
```
Processed 39546 nodes.
Error: 161 file(s) under /src/chromium/src/out are deps log dependencies of other targets, but have no generator target listing them as outputs.
Error: There are 156 missing dependency paths.
75 targets had depfile dependencies on 87 distinct generated inputs (from 34 rules) without a non-depfile dep path to the generator.
There might be build flakiness if any of the targets listed above are built alone, or not late enough, in a clean output directory.
```

There's one case that's arguably a false-positive: `Missing dep: gen/components/resources/about_credits.html uses build.ninja (generated by gn)` which may require a special case to not treat it as an error.

I also would like to have some sort of test for this, but don't really know where such a test (that needs an after-a-build state, basically) would hae tobe.

ilor commented:
Ping?

ilor commented:
Ping? I still believe this would be useful to have in ninja.

ilor commented:
Ping? This could help fix a lot of rare clean-build build flakes in the chrome gn/ninja build, see https://bugs.chromium.org/p/chromium/issues/detail?id=655123 for a list of issues found.

jcrben commented:
I'm just a random web developer wanting to try building chromium on macOS 10.12.6 using the instructions at https://chromium.googlesource.com/chromium/src/+/67.0.3360.2/docs/mac_build_instructions.md, and running into issues.

Hoping that any efforts to reduce flakiness / missing deps can be merged, so +1 on the ping.

When I ran this tool, I got 
![image](https://user-images.githubusercontent.com/5614134/36937246-40744146-1ec5-11e8-8431-78f0790a5aa9.png)

(note: this was not enough to help me fix my issue, which is apparently that mojom files are not generating their C++ files - chromium could perhaps use more instructions or more of a scripted setup)

bratell-at-opera commented:
I just ran into missing mojom.h files in a Chromium build and I wonder if this tool would have caught the problem when it was introduced. This seems like a good thing to have so I hope some maintainer can take a look.

atetubou commented on src/state.cc:
```diff
@@ -159,6 +159,23 @@ bool State::AddDefault(StringPiece path, string* err) {
+  bool found = false;
+  for (size_t i = 0; i < to->inputs_.size(); ++i) {
+    Edge* e = to->inputs_[i]->in_edge();
+    if (e && ((e == from) || PathExistsBetween(from, e))) {
```
```suggestion
    if (e && (e == from || PathExistsBetween(from, e))) {
```

ilor replied:
done (code moved to src/missing_deps.cc)

atetubou commented on src/ninja.cc:
```diff
@@ -505,6 +507,213 @@ int NinjaMain::ToolDeps(const Options* options, int argc, char** argv) {
+"options:\n"
+"  -g DIR   assume files under DIR are generated (default: .)\n"
+"    DIR is relative to the build directory, empty string disables\n"
+"    checking for files with a missing generator\n"
```
indent L690~L694

atetubou commented on src/ninja.cc:
```diff
@@ -505,6 +507,213 @@ int NinjaMain::ToolDeps(const Options* options, int argc, char** argv) {
+          generated_nodes_.insert(dep_nodes[i]);
+          printf("Missing dep: %s uses %s (generated by %s)\n",
+              node->path().c_str(), dep_nodes[i]->path().c_str(),
+              (*ne)->rule().name().c_str());
```
indent L630~L631

ilor replied:
obsolete after code move to src/missing_deps.cc

atetubou commented on src/ninja.cc:
```diff
@@ -505,6 +507,213 @@ int NinjaMain::ToolDeps(const Options* options, int argc, char** argv) {
+  string path = node->path();
+  char abs_path_buf[PATH_MAX];
+  char* rp = realpath(path.c_str(), abs_path_buf);
+  return (rp && (strlen(rp) > abs_gen_dir_.size())) &&
```
```suggestion
  return (rp && strlen(rp) > abs_gen_dir_.size()) &&
```

ilor replied:
obsolete after -g tool removal

atetubou commented on src/ninja.cc:
```diff
@@ -505,6 +507,213 @@ int NinjaMain::ToolDeps(const Options* options, int argc, char** argv) {
+  char abs_path_buf[PATH_MAX];
+  char* rp = realpath(path.c_str(), abs_path_buf);
+  return (rp && (strlen(rp) > abs_gen_dir_.size())) &&
+         (memcmp(rp, abs_gen_dir_.c_str(), abs_gen_dir_.size()) == 0);
```
```suggestion
         memcmp(rp, abs_gen_dir_.c_str(), abs_gen_dir_.size()) == 0;
```

ilor replied:
obsolete after -g tool removal

ilor commented:
Thanks for taking a look. What would be the preferred form of these tests?

My initial idea was that to show the behavior, I'd need several build directories with manually written ninja files for some types of mistakes. Then from some runner script I'd build in each of these and then run the tool to make sure I get the expected diagnostic. Would that be okay? I don't recall anything quite like it in the repo, so I'm not quite sure where I'd put it.

Something like this, assuming there's a generator for foo.h but foo.o doesn't dep on it:

    cd missing-deps-test1
    ninja clean  # assuming this is clean enough
    ninja foo.h   # must be manually built first -- full builds are flaky
    ninja foo.o
    ninja -t missingdeps   # should report foo.o lacks a dep to foo.h's generator

then repeated for a few other cases.

atetubou commented on src/state.cc:
```diff
@@ -159,6 +159,23 @@ bool State::AddDefault(StringPiece path, string* err) {
   return true;
 }
 
+bool State::PathExistsBetween(Edge* from, Edge* to) {
```
Better to detect cycle to prevent infinite recursion?

ilor replied:
Hm, is it possible to have cycles by this point? I thought we'd would error out before getting there. Needs checking in a test I guess.

ilor replied:
checked in a gtest that the code will error out before we get there

atetubou commented on src/ninja.cc:
```diff
@@ -505,6 +507,213 @@ int NinjaMain::ToolDeps(const Options* options, int argc, char** argv) {
+                                   gen_dir);
+  for (vector<Node*>::iterator it = nodes.begin(); it != nodes.end(); ++it) {
+    scanner.ProcessNode(*it);
+  }
```
Can we have gtest base test for MissingDependencyScanner?

ilor replied:
OK, I will have to see what I can come up with. Are there any existing tests that set up a build graph and depfile?

ilor replied:
gtests now included

jhasse commented:
> My initial idea was that to show the behavior, I'd need several build directories with manually written ninja files for some types of mistakes. Then from some runner script I'd build in each of these and then run the tool to make sure I get the expected diagnostic. Would that be okay? I don't recall anything quite like it in the repo, so I'm not quite sure where I'd put it.

Not sure if this is what you're looking for, but there's https://github.com/ninja-build/ninja/blob/master/misc/output_test.py which runs the ninja binary on a temporary file and checks its output.

ilor commented on src/ninja.cc:
```diff
@@ -514,6 +516,218 @@ int NinjaMain::ToolDeps(const Options* options, int argc, char** argv) {
+    abs_gen_dir_ = buf;
+    abs_gen_dir_.append("/");
+    abs_gen_dir_.append(gen_dir);
+    if (!realpath(abs_gen_dir_.c_str(), buf))
```
msvc/mingw don't like the realpath calls (AppVeyor build failed). Any recommendations?

jhasse replied:
You can use `std::filesystem::canonical`: https://en.cppreference.com/w/cpp/filesystem/canonical

You can use `#if _MSC_VER > 1910` to detect if its available and otherwise disable your tool (not really a problem, as our Windows release is build with Visual Studio 2017).

ilor replied:
Hm, not that I mind using C++17, but conditionally disabling the tool may be fairly ugly (no tools do that as far as I see?)

I noticed there's a Windows-specific AbsPath in src/includes_normalize.h, I can try moving it to src/util.h/cc so I can use it in the tool (may have to move the currently-local  InternalGetFullPathName as well to avoid changing too much other code), I think that'd be preferable over #including a weird header in a dependency analysis tool.

jhasse replied:
The browse tool is disabled on Windows and on any platform when `NINJA_PYTHON` isn't available at runtime.

ilor replied:
oh, thanks. I'll try it out and see how it goes, time permitting.

ilor replied:
this is now obsolete because the -g tool that required abspath is no longer part of this pull request to keep it more manageable

ilor commented:
Admittedly, the timing is not great with regards to me being able to spend as much time as I'd like on this. That said, I've rebased to master and fixed some minor issues, but will need a while for the tests.

The "main" obstacle is that the code uses classes like ImplicitDepLoader which want to load a real file and aren't triviallyfed test data; I guess I can use a temp file for now as I see some tests do that.

jonesmz commented:
@ilor Could you rebase this PR on the latest master branch?

ilor commented:
I'm working on this and should have something this week. It will be simplifies without the "files that lack a generator target" part (-g switch) because that adds a lot of complexity that's not needed to see the primary feature of the tool.

ilor commented:
Highlights from the update:
* Rebased onto master
* Separate cc/h pair for the tool, as it is large enough
* Unit test for the primary thing the tool should catch
* Remove -g switch and related code
* Avoid touching state.h/cc, keep the edge adjacency map

Removing -g and extracting to a separate source file simplified things a lot, and as a bonus now Windows should be able to build this without excessive ifdeffery or build config. I never found the -g tool to be very useful, it can be added later if there's a serious need.

Please take a look

jhasse commented on src/missing_deps.h:
```diff
@@ -0,0 +1,79 @@
+      return (from < other.from) || ((from == other.from) && (to < other.to));
+    }
+  };
+  // TODO: use a hash map here (~2x performance in a quick test)
```
Try:

```cpp
#if __cplusplus >= 201103L
  using EdgeAdjacencyMap = std::unordered_map<EdgePair, bool>;
#else
  typedef std::map<EdgePair, bool> EdgeAdjacencyMap;
#endif
```

ilor replied:
Will do. How do I configure the build to be in c++11 mode on Linux, by the way? I checked with a patched configure.py for now.

jhasse replied:
You can use the `CXXFLAGS` environment variable. Newer compilers default to C++11 mode btw.

ilor replied:
done

jhasse commented on src/missing_deps_test.cc:
```diff
@@ -0,0 +1,160 @@
+  CreateGraphDependencyBetween("compiled_object", "generated_header");
+  CreateGraphDependencyBetween("generated_header", "compiled_object");
+  // The missing-deps tool doesn't deal with cycles in the graph, beacuse
+  // there will be an error loading the graph befoer we get to the tool.
```
typo: befoer

ilor replied:
fixed in fixup

ilor commented on src/missing_deps.h:
```diff
@@ -60,8 +60,6 @@ namespace std {
-    // http://stackoverflow.com/a/1646913/126995
     uintptr_t uint_from = (uintptr_t)(k.from);
     uintptr_t uint_to = (uintptr_t)(k.to);
     return hash<uintptr_t>()(uint_from ^ (uint_to >> 3));
```
this is just a simplistic "hash" to get an unordered_map going, ended up plenty faster than map

jhasse commented on src/missing_deps.h:
```diff
@@ -0,0 +1,103 @@
+template <>
+struct hash<EdgePair> {
+  size_t operator()(const EdgePair& k) const {
+    uintptr_t uint_from = (uintptr_t)(k.from);
```
Please use `static_cast` instead of C-style cast. Or maybe the following will also work:
```cpp
uintptr_t uint_from{k.from};
```

ilor replied:
uinptr_t(k.from) works

ilor commented:
Thanks for the comments so far, should be all addressed. Please let me know what I can do next to get this into a merge-able state.

ilor commented:
Friendly ping

jhasse commented on src/graph.cc:
```diff
@@ -554,24 +554,26 @@ bool ImplicitDepLoader::LoadDepFile(Edge* edge, const string& path,
     if (!CanonicalizePath(const_cast<char*>(i->str_), &i->len_, &slash_bits,
                           err))
       return false;
-
```
why did you remove this line?

ilor replied:
no strong reason or strong opinion, I'll just put it back in

ilor commented:
Thanks for replying, I understand. While I think this can be very useful, it's not something I have to rush to any particular version.

I'll try to ping again in a week or a couple, maybe clean-up the branch history in the meantime (unless that's not welcome at this stage; I'm not sure how helpful/disruptive it'd be in the middle of a  pull request).

atetubou commented on src/missing_deps_test.cc:
```diff
@@ -0,0 +1,160 @@
+// Copyright 2011 Google Inc. All Rights Reserved.
```
2019?

ilor commented:
ping?

jonesmz commented on src/graph.cc:
```diff
@@ -507,9 +507,9 @@ bool ImplicitDepLoader::LoadDeps(Edge* edge, string* err) {
-                                    string* err) {
-  METRIC_RECORD("depfile load");
+bool ImplicitDepLoader::LoadDepfileDeps(Edge* edge, string path,
+                                        vector<Node*>* depfile_deps,
```
Consider using std:: for types from the standard library for newly added code.

ilor replied:
Thanks for pointing out the upcoming rule change!

jonesmz commented on src/graph.cc:
```diff
@@ -554,24 +554,27 @@ bool ImplicitDepLoader::LoadDepFile(Edge* edge, const string& path,
+  return true;
+}
 
+bool ImplicitDepLoader::LoadDepFile(Edge* edge, const string& path,
```
Is it practical for this string to instead be StringPiece?

ilor replied:
I don't think it would be useful, there is a grand total of two call-sites for this and both have a std::string instance at the ready.

jonesmz replied:
Alright.

jonesmz commented on src/graph.cc:
```diff
@@ -591,15 +594,18 @@ bool ImplicitDepLoader::LoadDepsFromLog(Edge* edge, string* err) {
+}
+
+void ImplicitDepLoader::ApplyLoadedDeps(Edge* edge, Node** nodes, int node_count) {
+  vector<Node*>::iterator implicit_dep = PreallocateSpace(edge, node_count);
```
Only now being made aware of the PreallocateSpace function, I think it's an abomination.

Can you instead use std::vector::reserve()?

I'd also argue that you should eliminate that function entirely, but if you don't I certainly will.

ilor replied:
As much as I don't fancy _either_ PreallocateSpace _or_ vector::reserve(), this looks a bit too close to the hot path to fiddle with as part of a larger change. I'd much rather see such changes done separately.

jonesmz replied:
Sure.

jonesmz commented on src/graph.cc:
```diff
@@ -554,24 +554,27 @@ bool ImplicitDepLoader::LoadDepFile(Edge* edge, const string& path,
+  if (!LoadDepfileDeps(edge, path, &depfile_deps, err))
+    return false;
+  if (!depfile_deps.empty())
+    ApplyLoadedDeps(edge, &depfile_deps[0], depfile_deps.size());
```
std::vector<>::data()

But would you consider taking an iterator pair instead of a pointer-to-pointer, and size_t?

ilor replied:
The other callsite uses deps->nodes, deps->node_count; DepsLog::Deps has a good ol' plain array of pointers.

ilor replied:
Also, vector::data() is c++11 :(

jonesmz replied:
It is? Did not know that. Thanks for the insights. 

All the more reason that Ninja should switch to C++11 fully, but *shrug*.

jhasse replied:
What's the advantage of std::vector<>::data()?

ilor replied:
Only that it's arguably nicer as &v[0] is unusual-looking.

jhasse replied:
Actually I just found out that there's another important difference: It's always okay to call `data()` on an empty vector. See 95e0cb5156d295e0510b5296e441eecb1f8b1806

jonesmz commented on src/graph.h:
```diff
@@ -247,6 +254,10 @@ struct ImplicitDepLoader {
 
+  /// Insert the loaded implicit dependencies (\a nodes) of \a edge
+  /// into the graph.
+  void ApplyLoadedDeps(Edge* edge, Node** nodes, int node_count);
```
node_count should be size_t, that's the representation for "number of objects in memory"

ilor replied:
DepsLog::Deps node_count is an int though, around src/deps_log.h:84.

jonesmz replied:
Alright. That's horrible, but you're right that it should be an int.

jonesmz commented on src/missing_deps.cc:
```diff
@@ -0,0 +1,160 @@
+
+void MissingDependencyPrinter::OnMissingDep(Node* node, const std::string& path,
+                                            const Rule& generator) {
+  printf("Missing dep: %s uses %s (generated by %s)\n", node->path().c_str(),
```
You might consider using iostreams, instead of printf. Printf is so.... ugh. For lack of better terminology to use.

ilor replied:
Fully agree on the terminology -- I used printf solely based on the fact that there are 0 (zero) usages of iostreams across ninja right now. If iostreams are "okay" (especially in a tool) then sure,

jonesmz replied:
While there's zero use of iostreams in terms of `std::cout`, there are several places where std::stringstream is used to construct a string (which, in a few situations, is then passed to fprintf...?)

While I am not an official project maintainer, and therefore one of the official maintainers can just overrule me, personally I think (for all it's warts...) iostreams is miles and miles ahead of fprintf, not the least reason of which is that it doesn't explicitly require nul-terminated strings.

ilor replied:
Let's see how it goes with std::cout.

jonesmz commented on src/missing_deps.cc:
```diff
@@ -0,0 +1,160 @@
+    ImplicitDepLoader dep_loader(state_, deps_log_, disk_interface_,
+                                 &parser_opts);
+    vector<Node*> depfile_deps;
+    string depfile = edge->GetUnescapedDepfile();
```
consider `std::` for newly added code.

jonesmz commented:
Given the upcoming rule change (https://github.com/ninja-build/ninja/pull/1616/files):

> * Any code merged into the Ninja codebase which will be part of the main
>   executable must compile as C++03. You may use C++11 features in a test or an
>   unimportant tool if you guard your code with `#if __cplusplus >= 201103L`.

You might consider adopting C++11 entirely for this new tool. It would allow for things like ranged based for loops, which would benefit several places in your newly added code.

jonesmz commented on src/missing_deps.cc:
```diff
@@ -0,0 +1,160 @@
+  }
+
+  if (!missing_deps.empty()) {
+    set<string> missing_deps_rule_names;
```
Could you use StringPiece here, instead of string? Saves allocations.

ilor replied:
It's somewhat awkward. This StringPiece doesn't work in a set since it has no operator<, and there's no hash_set typedef either.

jonesmz replied:
Makes sense.

jonesmz commented on src/missing_deps.cc:
```diff
@@ -0,0 +1,160 @@
+}
+
+bool MissingDependencyScanner::PathExistsBetween(Edge* from, Edge* to) {
+  EdgePair key;
```
Why not provide a constructor for EdgePair, so you can formulate this as

EdgePair key(from, to);

?

jonesmz commented on src/missing_deps.h:
```diff
@@ -0,0 +1,103 @@
+
+ private:
+#if __cplusplus >= 201103L
+  using EdgeAdjacencyMap = std::unordered_map<EdgePair, bool>;
```
Wouldn't the hash_map implementation found in hash_map.h be sufficient for this?

ilor replied:
As far as I can tell no, it seems only useful for maps keyed by a StringPiece.

jonesmz replied:
Ah, yes, you're correct. my mistake.

jonesmz commented on src/ninja.cc:
```diff
@@ -915,7 +940,7 @@ const Tool* ChooseTool(const string& tool_name) {
     for (const Tool* tool = &kTools[0]; tool->name; ++tool) {
       if (tool->desc)
-        printf("%10s  %s\n", tool->name, tool->desc);
+        printf("%11s  %s\n", tool->name, tool->desc);
```
Could you elaborate on this change?

ilor replied:
This is to align the list of tools properly -- the "missingdeps" line stands out otherwise with its short description. That's because "missingdeps" happens to be longer than the other tool names.
Roughly:
```
     browse  browse dependency graph in a web browser
      clean  clean built files
   commands  list all commands required to rebuild given targets
       deps  show dependencies stored in the deps log
missingdeps  check deps log dependencies on generated files
```
vs
```
    browse  browse dependency graph in a web browser
     clean  clean built files
  commands  list all commands required to rebuild given targets
      deps  show dependencies stored in the deps log
missingdeps  check deps log dependencies on generated files
```

jonesmz replied:
Sounds good.

jonesmz commented on src/missing_deps.h:
```diff
@@ -45,14 +45,15 @@ class MissingDependencyPrinter : public MissingDependencyScannerDelegate {
 struct EdgePair {
-  Edge* from;
-  Edge* to;
+  EdgePair(Edge* from_, Edge* to_) : from(from_), to(to_) {}
```
I'm not trying to nitpick you, sorry :(

I think that the typical Ninja coding conventions have the underscore for member variables, and the constructor parameters are underscore-less.

ilor replied:
I don't mind at all

jhasse replied:
btw: There's no need to use a different name,  `EdgePair(Edge* from, Edge* to) : from(from), to(to) {}` works.

ilor replied:
yeah, I'm aware, but I really don't like relying on this particular nook of c++ name scoping

jonesmz commented on src/missing_deps.cc:
```diff
@@ -0,0 +1,158 @@
+// Copyright 2019 Google Inc. All Rights Reserved.
```
IANAL -- but I am very familiar with copyright issues, having worked with them plenty over my career.

Please consider carefully the license header, unless you are a Google employee.

While the Ninja project does have the explicit guideline that all files have "Copyright *year* Google Inc.", this is very much not the correct thing to do for non-Google employees.

If you are not a Google employee, then you are the owner of the copyright for newly written code. In such case, you can't simply say "Google owns this" without some explicit agreement with them to have the copyright transfer.

jonesmz replied:
Allow me to clarify my poor wording.

*obviously* you can *say* whatever you want.

I meant "Saying that Google holds the copyright on this code is not factually correct simply because the file says so. The copyright is retained by the actual author in non-"work for hire" situations unless there's an explicit agreement between the author and Google saying that the rights have been transferred".

ilor replied:
I believe this would be more appropriate to discuss in https://github.com/ninja-build/ninja/pull/1616 that explicitly mentions the licensing, vs in a random change that happens to add a file.

And anyway, the change here is from a corporate email address with a google CLA in place already due to contributions elsewhere; and https://opensource.google.com/docs/releasing/contributions/#copyright doesn't indicate worrying about the copy header is required.

jonesmz replied:
No need for you to do anything. I only wanted to make sure you were aware.

ilor commented:
Friendly/festive ping

jhasse commented:
Please have a look at the conflict in src/graph.cc.

ilor commented:
Thanks for pointing that out. Rebased on top of master to fix the conflict an cleaned up branch history while at it. Then noticed there's a cmake build now too, which the last fixup deals with.

jhasse commented:
It's really time for 1.10 to be released and I think we already included lots of useful stuff, which needs testing.

Therefore I'm assigning this PR to 1.11, hope you're okay with that. I just didn't have the time to test this thoroughly and it includes some non-tool-only changes ;)

ilor commented:
That's fine. I assume I should wait until the 1.10 release and then rebase the docs then so they say ... since 1.11 etc.?

jhasse commented:
Yes. No need to wait for the release for updating the docs though.

ilor commented:
Apologies for the delas. Rebased with no conflicts, marked doc to indicate 1.11 as the version.

ilor commented:
Friendly ping

WhiteHalmos commented:
This PR has been super useful whenever I change dependencies (I compiled ninja with this patch). Looking forward to this being merged into master.

ilor commented:
ping

ilor commented:
Ping

deepakcse87 commented:
Nice tool. Looking forward to use it when it gets merged into master.

ilor commented:
@jhasse friendly ping

jhasse commented:
This PR is on the 1.11.0 milestone, I will definitely have a look at it before doing that release. Currently I'm working on 1.10.2 though, please be patient.

ilor commented:
That's good to hear, I'll be waitng.

jhasse commented:
Can you rebase on master? I think I found a crash in the unit test due to an assert.

ilor commented:
The test crash seems to be because the test code that writes a depfile needs to be adjusted after cc79afbc0510b7320d35fb481f0f200dec92cb43 changed some bits in deps_log.cc

jhasse commented on src/missing_deps.h:
```diff
@@ -94,12 +68,14 @@ struct MissingDependencyScanner {
  private:
-#if __cplusplus >= 201103L
-  using EdgeAdjacencyMap = std::unordered_map<EdgePair, bool>;
+#if (__cplusplus >= 201103L) || (_MSC_VER >= 1900)
```
`_MSC_VER >= 1900` shouldn't be needed as MSVC sets `__cplusplus` correctly with `/Zc:__cplusplus`.

ilor replied:
ack

ilor commented:
ping -- this should be better now, the macos failure looks like a infra issue

jhasse reviewed (changes requested):
Thanks! The code looks good :) There's one performance impact though:

While `implicit_dep` is still preallocated, the new temporary vector `depfile_deps` will grow on the fly. It's also now two loops because of the temporary vector.

Either

1. show that this has no significant performance impact with an appropriate benchmark,
2. or pass a callback to `LoadDepfileDeps` instead of a pointer to an vector. Call it and pass the node instead of doing `depfile_deps->push_back(node);`.

ilor commented:
@jhasse thanks, with all the quaint pre-C++11 support going on "use a callback" is a bit less enticing than it should otherwise be but I'll probably end up doing that. I may follow up with some perf experiments later if I find the time because I do feel like one extra vector and some pointer copies there should be immaterial.

jhasse commented:
I hope it's okay I've tried the "Allow changes from maintainers" feature of GitHub and pushed to your branch 🙈 

It isn't pretty, especially `PreallocateSpace` and `CreatePhonyInEdge` shouldn't be public. Maybe you have a better idea.

ilor commented:
I think the virtual function approach works out better now, especially when viewed vs the base branch. Should be little effectively no changes on the main code path now.

ilor commented:
friendly ping

jhasse commented on src/graph.cc:
```diff
@@ -525,9 +525,14 @@ struct matches {
-bool ImplicitDepLoader::LoadDepFile(Edge* edge, const string& path,
-                                    string* err) {
+bool ImplicitDepLoader::LoadDepFile(Edge* edge, const std::string& path,
+                                    std::string* err) {
```
please remove `std::` in both cases to show that this function's signature hasn't changed.

ilor replied:
done, a lot less diff all over graph.h/cc now

jhasse commented on src/graph.cc:
```diff
@@ -660,3 +670,26 @@ void ImplicitDepLoader::CreatePhonyInEdge(Node* node) {
+                                                std::string* err) {
+  dep_nodes_ = dep_nodes;
+  bool ret = LoadDepfileDeps(edge, path, err);
+  dep_nodes_ = NULL;
```
What if dep_nodes_ wasn't `NULL` before?

jhasse replied:
Okay that can't happen. But why even set it to `NULL` here?

ilor replied:
I did this to avoid holding on to a potentially-dangling pointer. This is admittedly a bit dodgy if someone were to call LoadDeps on this loader afterwards. I'm considering removing LoadDepfileDepNodes and adding a direct setter for dep_nodes_, so the calling code would be
```
dep_loader.SetDepNodesOutput(&depfile_deps);
dep_loader.LoadDeps(edge, depfile, &err);
```
instead of the current
```
dep_loader.LoadDepfileDepNodes(edge, depfile, &depfile_deps, &err);
```
Or maybe even just take the vector pointer in the constructor of the custom dep loader to keep things more straightforward since the pointer can then be easily shown as something necessary. That seems the way to go.

jhasse replied:
> Or maybe even just take the vector pointer in the constructor of the custom dep loader to keep things more straightforward since the pointer can then be easily shown as something necessary.

+1

ilor replied:
PTAL now

jhasse commented on src/graph.h:
```diff
@@ -271,6 +282,30 @@ struct ImplicitDepLoader {
 
+/// ImplicitDepLoader variant that returns dep nodes without updating graph
+/// deps.
+struct ImplicitDepNodeLoader : public ImplicitDepLoader {
```
Move the whole class to missing_deps.h/cc?

jhasse commented on src/missing_deps_test.cc:
```diff
@@ -0,0 +1,162 @@
+  CreateGraphDependencyBetween("generated_header", "compiled_object");
+  // The missing-deps tool doesn't deal with cycles in the graph, beacuse
+  // there will be an error loading the graph before we get to the tool.
+  // This test is to illstrate that.
```
typo: illustrate

jhasse reviewed (commented):
That's a good solution :)
